arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -0,0 +1,438 @@
+	}
+	return &BatcherAddrMonitor{
+		initAddresses:     initAddresses,
+		db:                db,
+		l1Reader:          l1Reader,
+		seqInboxAddr:      seqInboxAddr,
+		seqInboxInterface: seqInboxInterface,
+		deployAt:          deployAt,
+		parentHeight:      deployAt,
+	}
+}
+
+func (b *BatcherAddrMonitor) AddBatchPosterSetEvents(events []BatcherAddrEvent) error {
+	if len(events) == 0 {
+		return nil
+	}
+	b.events = append(b.events, events...)
+	// Sort events by l1Height to ensure correct processing order.
+	// Since BatcherAddr events are infrequent, the performance impact of sorting is negligible.
+	sort.Slice(b.events, func(i, j int) bool {
+		return b.events[i].L1Height < b.events[j].L1Height
+	})
+	b.lastEventL1Height = b.events[len(b.events)-1].L1Height
+	b.cached = false
+	return b.Store()
+}
+
+func (b *BatcherAddrMonitor) GetValidAddresses(l1 uint64) []common.Address {
+	if l1 > b.l1Height {
+		// If the target L1 height is greater than the latest known L1 height,
+		// return an empty slice. The caller should wait until the monitor has
+		// observed at least this L1 height before calling this function.
+		return []common.Address{}
+	}
+
+	if len(b.events) == 0 || b.events[0].L1Height > l1 {
+		return b.initAddresses
+	}
+
+	// If the target L1 height is within the latest cached window, return the cached result.
+	// In a practical scenario, this is the most common case. Here means that during the time
+	// from `lastEventL1Height` to `l1Height`, the `events` are not changed. It is not needed to
+	// calculate valid batcher addresses.
+	latestCachedWindow := l1 >= b.lastEventL1Height && l1 <= b.l1Height
+	if b.cached && latestCachedWindow {
+		return b.cachedAddresses
+	}
+
+	result := map[common.Address]bool{}
+	for _, addr := range b.initAddresses {
+		result[addr] = true
+	}
+
+	for _, event := range b.events {
+		if event.L1Height > l1 {
+			break
+		}
+
+		result[event.Addr] = event.IsBatcher
+	}
+
+	var validAddrs []common.Address
+	for addr, isBatcher := range result {
+		if isBatcher {
+			validAddrs = append(validAddrs, addr)
+		}
+	}
+
+	if latestCachedWindow {
+		b.cached = true
+		b.cachedAddresses = validAddrs
+	}
+
+	return validAddrs
+}
+
+func (b *BatcherAddrMonitor) SetParentHeight(height uint64) {
+	b.parentHeight = height
+}
+
+func (b *BatcherAddrMonitor) SetL1Height(height uint64) {
+	b.l1Height = height
+}
+
+func (b *BatcherAddrMonitor) GetConfirmedParentHeight() uint64 {
+	return b.parentHeight
+}
+
+func (b *BatcherAddrMonitor) LookupEvents(ctx context.Context, fromBlock, toBlock uint64) ([]BatcherAddrEvent, error) {
+	from := big.NewInt(0).SetUint64(fromBlock)
+	to := big.NewInt(0).SetUint64(toBlock)
+	query := ethereum.FilterQuery{
+		BlockHash: nil,
+		FromBlock: from,
+		ToBlock:   to,
+		Addresses: []common.Address{b.seqInboxAddr},
+		Topics: [][]common.Hash{
+			{ownerFunctionCalledID},
+			{common.BigToHash(big.NewInt(1))},
+		},
+	}
+	logs, err := b.l1Reader.Client().FilterLogs(ctx, query)
```

Comment:
what happens if the node goes down and it misses the logs? How do we recover?

Reply:
We store the `batcher update events` and the last processed parent height. And restore them from db every time it starts